design_notes/XDN03 - xlang Type System.md

```diff
@@ -15,3 +15,142 @@ status: draft
 This design note describes the xlang's [type system](https://en.wikipedia.org/wiki/Type_system).
 The type system is designed to interoperate with a variety of object-oriented programming
 languages in a natural and familiar way.
+
+## Namespaces
+
+A namespace is a naming scope used to organize code and avoid naming collisions. All type
+categories in the xlang type system (enums, structs, delegates, interfaces, and runtime classes)
+except for fundamental types live in a namespace.
+
+Namespaces may contain other namespaces.
+
+xlang namespaces and type names are case preserving but insensitive. This means that you cannot
+have namespaces or type names that vary only by case. For example: you cannot have `Foo.SomeType`
+and `foo.SomeType` nor can you have `Foo.SomeType` and `Foo.someType`. However, you can have
+`Foo.SomeType` and `foo.AnotherType`. In this case `SomeType` and `AnotherType` would be considered
+as belonging to the same namespace.
+
+## Fundamental Types
+
+The xlang Type system includes a core set of built-in primitive types.
+
+xlang Type | Type Description
+---------- | ----------------------------------------------------------
+Int16      | a 16 bit signed integer
+Int32      | a 32 bit signed integer
+Int64      | a 64 bit signed integer
+UInt8      | an 8 bit unsigned integer
+UInt16     | a 16 bit unsigned integer
+UInt32     | a 32 bit unsigned integer
+UInt64     | a 64 bit unsigned integer
+Single     | a 32-bit IEEE 754 floating point number
+Double     | a 64-bit IEEE 754 floating point number
+Char16     | a 16-bit non-numeric value representing a UTF-16 code unit
+Boolean    | an 8-bit Boolean value
+String     | an immutable sequence of Char16s used to represent text
+Guid       | A 128-bit standard [universally unique identifier](https://en.wikipedia.org/wiki/Universally_unique_identifier)
+Object     | An xlang object of unknown type
+
+> Note: string encoding for xlang is currently under review. In particular, it has not been decided
+if xlang will support UTF-8 strings in some capacity. A design note will be commissioned for this
+decision.
+
+## Enums
+
+An enum type is a distinct value type with a set of named value. Each named value in an enum
+corresponds to an constant integer value.
+
+Each enum type has an underlying integral type. The only legal underlying integral types for enums
+in xlang are Int32 and UInt32.
+
+An enum type with an underlying integral type of UInt32 is considered a Flags enum. Flags enums
```

Comment:
We definately need unsigned integers. We had lots of debate about that back in the WinRT days.

Some languages, e.g., JavaScript, will simply convert them to a larger signed numeric type. Int64, signed or unsigned, is a problem for JavaScript because it intrinsicly doesn't support that large of a numeric value.

Reply:
I'm not sure if Brent is trying to reply to me or this thread but to add some context to my comment, I definitely think unsigned is important, but I'm concerned about exactly the boundaries you describe. In Rome, I forget where exactly, but we ran into the widening issue that translating a uint64 into java was unwieldy at best (there is a BigInteger class but that feels like a hack). We got away with it because it wasn't intended for arithmetic operations so the sign didn't actually matter (i think it was an id maybe) but I want to make sure that we don't promise something at the type system level that is going to project poorly (or at least make it easy to know at author time that it may not work the best)

Reply:
We don't want the type system to artificially limit API authors. An API author designing an API for Java or JavaScript may naturally avoid certain types when his target audience is JavaScript or some other limited language, but we should not restrict the type system as a whole to the lowest common denominator.

Reply:
We ran into the [U]Int64 & JavaScript issue back in the Win8 days. There are APIs that absolutely require 64-bit integers (e.g. file system APIs). We cannot cripple such APIs by lowering the type system to the least common denominator of all languages.

Yes, this means that some APIs will potentially by unusable (or usable with restrictions). Better than the alternative.

Reply:
Yeah I understand the desire for the flexibility and I think we are ok as long as there is an annotation or warning or something that we can use in metadata to say "not projectable everywhere" or something. Just want to call it out so we don't inadvertently lead devs down a path they can't get out of without breaking changes. 

Assuming that an author is thinking about *all* target languages at the time of authoring the API may be a bit ambitious as projects evolve etc. (For instance Rome would love to support JavaScript someday but that was not on the mind of the api developers at the very beginning of our cross plat effort)

Reply:
FYI, since this thread has jumped topics, I started a [new conversation](#r247248656) to address @jonwis and @kennykerr  comments about enums.

Reply:
I'm going to leave this conversation open for now, but I'm basically with @BrentRe that we need unsigned and 64 bit ints, even though various languages may have issue with them. @bbowman In Java's case, aren't there a bunch of helper methods for dealing with unsigned ints?

Reply:
There are some helpers like https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html#divideUnsigned-int-int- but it requires the consumer of the value to understand its unigned and never treat it otherwise. There is also the BigInteger class that can go beyond 64bit but also doesn't "feel" like a primitive does. 

I still want to hear how we handle the casewhich you say "even though various languages may have issue with them" Perhaps there needs to separate design note on language inconsistencies and how that appears in metadata / the authoring experience.

Somewhat tangentially, will there be any annotation support at the metadata level?

Reply:
@bbowman can you clarify what you mean by annotation support at the metadata level?

Reply:
I think its important that there is a communication channel to both xlang component authors and xlang component consumers for information that cannot be perfectly encoded in the API signature either due to deficiencies of the type system (i.e. Uint64 can't be perfectly projected to some languages) or due to some authoring decision (i.e. this parameter can't be null or only values 1-10 will be accepted). 

Obviously comments are one form pf this communication, however, due to the complexities of spoken languages and the fact that comments are traditionally "hands off", a more structured approach in metadata might be useful to "mark" certain attributes on the api such that the projection can inherit them to inform the component consumer and the tooling can add its own to help inform the component author. To my knowledge the WinRT tooling already allows for some "attributes" being added to APIs. I'm just suggesting we have a design note that formalizes what this looks like, how authors can expand it, etc.

Inspiration comes from Java: https://developer.android.com/reference/android/support/annotation/IntRange#annotations

Reply:
(Also a Design Note on API comments and tooling is probably worthwhile. That seems like an interesting problem in its own)

Reply:
yes, it does sound like an interesting problem space - can you file and issue @bbowman? Or you can start work on an XDN - anyone in the community can [author design notes](https://github.com/Microsoft/xlang/blob/master/design_notes/XDN00%20-%20Design%20Note%20Process.md#2-the-process)

Reply:
THis is now tracked by issue #130